pkg/controllers/provisioningv2/rke2/rkecluster/controller.go

```diff
@@ -37,38 +38,44 @@ func Register(ctx context.Context, clients *wrangler.Context) {
 	}, clients.RKE.RKECluster(), clients.RKE.RKEControlPlane())
 }
 
-func (h *handler) UpdateSpec(_ string, cluster *v1.RKECluster) (*v1.RKECluster, error) {
+func (h *handler) OnChange(_ string, cluster *v1.RKECluster) (*v1.RKECluster, error) {
 	if cluster == nil {
 		return nil, nil
 	}
 
-	if cluster.Spec.ControlPlaneEndpoint == nil {
+	capiCluster, err := rke2.GetOwnerCAPICluster(cluster, h.capiClusterCache)
+	if apierrors.IsNotFound(err) {
+		logrus.Debugf("[rkecluster] %s/%s: waiting: CAPI cluster does not exist", cluster.Namespace, cluster.Name)
+		h.rkeCluster.EnqueueAfter(cluster.Namespace, cluster.Name, 10*time.Second)
+		return cluster, generic.ErrSkip
+	}
+	if err != nil {
+		logrus.Errorf("[rkecluster] %s/%s: error getting CAPI cluster %v", cluster.Namespace, cluster.Name, err)
+		return cluster, err
+	}
+
+	if capiannotations.IsPaused(capiCluster, cluster) {
+		logrus.Infof("[rkecluster] %s/%s: waiting: CAPI cluster or RKECluster is paused", cluster.Namespace, cluster.Name)
+		h.rkeCluster.EnqueueAfter(cluster.Namespace, cluster.Name, 10*time.Second)
```

Comment:
Do we really need to re-enqueue after 10 seconds here? This seems rudimentary compared to the tools we have available. We could use related resource to watch for the capi cluster to change...?

Reply:
This is what I came up with:
```
	relatedresource.Watch(ctx, "rke-cluster-trigger", func(namespace, name string, obj runtime.Object) ([]relatedresource.Key, error) {
		if capiCluster, ok := obj.(*capi.Cluster); ok && !capiCluster.Spec.Paused {
			return []relatedresource.Key{{
				Namespace: namespace,
				Name:      name,
			}}, nil
		}
		return nil, nil
	}, clients.RKE.RKECluster(), clients.CAPI.Cluster())
```
though I'm not sure this encapsulates what we want either, as we don't really need to enqueue the cluster/bootstrap/machine objects _while_ the CAPI cluster is unpaused, we just want to pause reconcilation _until_ the cluster is unpaused, which may require some wrangler updates.

Reply:
I don't see an issue if the controller reconciles the object based on related resource if the controller isn't going to make a lot of noise/be detrimental to the performance of rancher.

Reply:
It shouldn't affect performance, there will be a few extra reconcilations done until the PR that restructures the planner around these predicates goes in, since the CAPI cluster will continuously update the `status.InfrastructureReady` field and a few conditions, but even that is currently negligible. I'll go ahead and update.